LocationManager/INTULocationManager/INTULocationManager.m

```diff
@@ -692,6 +758,110 @@ - (INTULocationAccuracy)achievedAccuracyForLocation:(CLLocation *)location
     }
 }
 
+#pragma mark Internal heading methods
+
+/**
+ Adds the given heading request to the array of requests and starts heading updates.
+ */
+- (void)addHeadingRequest:(INTUHeadingRequest *)headingRequest
+{
+    NSAssert(headingRequest, @"Must pass in a non-nil heading request.");
+
+    // If heading services are not available, just return
+    if ([INTULocationManager headingServicesState] == INTUHeadingServicesStateUnavailable) {
+        if (headingRequest.block) {
+            headingRequest.block(nil, INTUHeadingStatusUnavailable);
+        }
+        INTULMLog(@"Heading Request (ID %ld) NOT added since device heading is unavailable.", (long)headingRequest.requestID);
+        return;
+    }
+
+    __INTU_GENERICS(NSMutableArray, INTUHeadingRequest *) *newHeadingRequests = [NSMutableArray arrayWithArray:self.headingRequests];
+    [newHeadingRequests addObject:headingRequest];
+    self.headingRequests = newHeadingRequests;
+    INTULMLog(@"Heading Request added with ID: %ld", (long)headingRequest.requestID);
+
+    if (!self.isUpdatingHeading) {
+        self.isUpdatingHeading = YES;
+
+        [self.locationManager startUpdatingHeading];
+    }
+}
+
+/**
+ Removes a given heading request from the array of requests and stops heading updates if needed.
+ */
+- (void)removeHeadingRequest:(INTUHeadingRequest *)headingRequest
+{
+    __INTU_GENERICS(NSMutableArray, INTUHeadingRequest *) *newHeadingRequests = [NSMutableArray arrayWithArray:self.headingRequests];
+    [newHeadingRequests removeObject:headingRequest];
+    self.headingRequests = newHeadingRequests;
+
+    [self stopUpdatingHeadingIfPossible];
+}
+
+/**
+ Checks to see if there are any outstanding headingRequests, and if there are none, informs CLLocationManager to stop sending
+ heading updates. This is done as soon as heading updates are no longer needed in order to conserve the device's battery.
+ */
+- (void)stopUpdatingHeadingIfPossible
+{
+    if (self.headingRequests.count == 0) {
+        [self.locationManager stopUpdatingHeading];
+        if (self.isUpdatingHeading) {
+            INTULMLog(@"Location services heading updates have stopped.");
+        }
+        self.isUpdatingHeading = NO;
+    }
+}
+
+/**
+ Iterates over the array of active heading requests and processes each
+ */
+- (void)processRecurringHeadingRequests
+{
+    for (INTUHeadingRequest *headingRequest in self.headingRequests) {
+        [self processRecurringHeadingRequest:headingRequest];
+    }
+}
+
+/**
+ Handles calling a recurring heading request's block with the current heading.
+ */
+- (void)processRecurringHeadingRequest:(INTUHeadingRequest *)headingRequest
```

Comment:
Yeah, this implementation is one option. I was just asking what you thought about returning some sort of error status for this case instead, because unlike for location, there is no specific callback from CLLocationManager when there is a heading update error. So if we want to notify clients of errors, we'll have to do it by noticing when we receive an invalid heading and taking appropriate action at that time, instead of just discarding invalid headings.